src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -211,57 +210,75 @@ public ProgrammingExerciseService(ProgrammingExerciseRepository programmingExerc
      * @throws GitAPIException If something during the communication with the remote Git repository went wrong
      * @throws IOException     If the template files couldn't be read
      */
-    @Transactional // TODO: apply the transaction on a smaller scope
-    // ok because we create many objects in a rather complex way and need a rollback in case of exceptions
     public ProgrammingExercise createProgrammingExercise(ProgrammingExercise programmingExercise, boolean isImportedFromFile) throws GitAPIException, IOException {
-        programmingExercise.generateAndSetProjectKey();
         final User exerciseCreator = userRepository.getUser();
-
         VersionControlService versionControl = versionControlService.orElseThrow();
+
+        programmingExercise.setSolutionParticipation(null);
+        programmingExercise.setTemplateParticipation(null);
```

Comment:
Why is this needed here? Please document this behavior (or find a better solution ^^)

Reply:
Currently, the client sends a solution and template participation object (filled with null values) when creating a programming exercise.
In either way (saving the exercise or the participation object before the other one), this leads to an exception at runtime.
As the participations objects are just dummy values representing the data structure in the client, we set this to `null`.

Reply:
Please add some inline code documentation. I agree that it looks strange and other developers might stumble upon it in the future